perf/light.cc

```diff
@@ -0,0 +1,653 @@
+	assert(p_extents_count == &extents_count);
+	++*p_extents_count;
+	return malloc(light_extent_size);
+}
+
+inline void
+light_free_extend(void *ctx, void *p)
+{
+	size_t *p_extents_count = (size_t *) ctx;
+	assert(p_extents_count == &extents_count);
+	--*p_extents_count;
+	free(p);
+}
+
+////////////////////////////// Light Fixture ///////////////////////////////////////////////////////////////////////////
+
+class LightFixture : public benchmark::Fixture {
+public:
+	void
+	SetUp(const benchmark::State &state)
+	{
+		ht = static_cast<struct light_core *>(malloc(sizeof(*ht)));
+		light_create(ht, light_extent_size, light_malloc_extend,
+			     light_free_extend, &extents_count, 0);
+	}
+
+	void
+	TearDown(const benchmark::State &state)
+	{
+		light_destroy(ht);
+		free(ht);
+		ht = nullptr;
+	}
+protected:
+
+	void
+	Fill(TupleHolder &holder, std::size_t count)
+	{
+		for (auto t = holder.tuples.begin(); t != holder.tuples.begin() + count; ++t) {
+			light_insert(ht, t->hash, t->raw);
+		}
+	}
+
+	void
+	Erase(TupleHolder &holder, std::size_t count)
+	{
+		for (auto t = holder.tuples.begin(); t != holder.tuples.begin() + count; ++t) {
+			light_delete_value(ht, t->hash, t->raw);
+		}
+	}
+
+	struct light_core *ht;
+};
+
+// Insert random values into light; no warm up - hashtable is empty at the benchmark start.
+BENCHMARK_DEFINE_F(LightFixture, InsertRandValue)
+(benchmark::State& state)
+{
+	std::int64_t insertion_count = 0;
+	for (auto s : state) {
+		state.PauseTiming();
+		// state.range(0) stores current range argument.
+		TupleHolder data(state.range(0));
+		state.ResumeTiming();
+
+		for (auto &v : data.tuples) {
+			benchmark::DoNotOptimize(light_insert(ht, v.hash, v.raw));
+			insertion_count++;
+		}
+	}
+	state.SetItemsProcessed(insertion_count);
+}
+
+// Lookup random values; value is supposed to be presented in hashtable (no misses).
+BENCHMARK_DEFINE_F(LightFixture, FindRandValue)
+(benchmark::State& state)
+{
+	std::int64_t lookup_count = 0;
+	for (auto s : state) {
+		state.PauseTiming();
+		TupleHolder data(state.range(0));
+		Fill(data, state.range(0));
+		data.shuffle(state.range(0));
+		state.ResumeTiming();
+
+		for (auto &v : data.tuples) {
+			benchmark::DoNotOptimize(light_find(ht, v.hash, v.raw));
+			lookup_count++;
+		}
+	}
+	state.SetItemsProcessed(lookup_count);
+}
+
+// Lookup random values (most of them should not be presented in the table).
+BENCHMARK_DEFINE_F(LightFixture, FindRandValueWithMisses)
+(benchmark::State& state)
+{
+	std::int64_t lookup_count = 0;
+	for (auto s : state) {
+		state.PauseTiming();
+		TupleHolder data(state.range(0));
+		Fill(data, state.range(0));
+		// Re-generate dataset so it contains random values.
+		TupleHolder missing_values(state.range(0));
+		state.ResumeTiming();
+
+		for (auto &v : missing_values.tuples) {
+			benchmark::DoNotOptimize(light_find(ht, v.hash, v.raw));
+			lookup_count++;
+		}
+	}
+	state.SetItemsProcessed(lookup_count);
+}
+
+// Lookup random values by key.
```

Comment:
Why do you have different sets of benchmarks for key- and value- wise lookups? Please also reflect this in your results: it's not obvious for which type of lookup are the results presented.

Reply:
What do you mean by that? I've underlined that for each test dataset is generated from scratch:

> Before each test we prepare vector of tuples storing truly random values.

Reply:
I mean you have two different benchmarks for looking up by value and only one for looking up by key.

Reply:
Yep, but I considered it to be insignificant so skipped.